internal/migrations/changeset_listener.go

```diff
@@ -0,0 +1,272 @@
+)
+
+// Changeset Extension polls the changesets from the old chain  during migration.
+// The changesets received are broadcasted to the eventstore under ChangesetMigrationEvent type for voting.
+// The changesets are then applied to the new chain's database once the resolution is approved.
+const (
+	ListenerName = "migrations"
+	retryDelay   = 1 * time.Second // polling frequency for changesets
+)
+
+func init() {
+	// Register the listener with the name "migrations"
+	err := listeners.RegisterListener(ListenerName, Start)
+	if err != nil {
+		panic(err)
+	}
+}
+
+type MigrationConfig struct {
+	// StartHeight is the block height at which the migration started on the old chain.
+	StartHeight int64
+
+	// End height is the block height at which the migration ended on the old chain.
+	EndHeight int64
+
+	// AdminListenAddress is the address of the admin server to receive changesets from.
+	AdminListenAddress string
+
+	// AdminPass is the admin server password.
+	AdminPass string
+
+	// KwildTLSCertFile is the path to the TLS certificate file for the Kwil node.
+	// The path should be an absolute path or relative to the node's root directory.
+	KwildTLSCertFile string
+}
+
+func Start(ctx context.Context, service *common.Service, eventStore listeners.EventStore) error {
+	// Get the migration config from the service
+	cfg := &MigrationConfig{}
+	migrationCfg, ok := service.ExtensionConfigs[ListenerName]
+	if !ok {
+		service.Logger.Warn("no migration config provided, skipping migration listener")
+		return nil // no migration config, nothing to do
+	}
+
+	err := cfg.ExtractConfig(migrationCfg)
+	if err != nil {
+		return err
+	}
+
+	var dialOpt []adminclient.Opt
+	if cfg.KwildTLSCertFile != "" {
+		dialOpt = append(dialOpt, adminclient.WithTLS(cfg.KwildTLSCertFile, "", ""))
+	}
+
+	if cfg.AdminPass != "" {
+		dialOpt = append(dialOpt, adminclient.WithPass(cfg.AdminPass))
+	}
+
+	// Admin RPC client connection to the old chain
+	adminClient, err := adminclient.NewClient(ctx, cfg.AdminListenAddress, dialOpt...)
+	if err != nil {
+		return err
+	}
+
+	// Get the block height of the last changesets received from the old chain
+	// If no height is found, start from the start height of the migrations.
+	// If all the changesets have already been received, the listener will exit.
+	var currentHeight, lastHeight int64
+	lastHeight, err = getLastStoredHeight(ctx, eventStore)
+	if err != nil {
+		return err
+	}
+
+	if lastHeight == 0 {
+		currentHeight = cfg.StartHeight
+	} else if lastHeight >= cfg.EndHeight {
+		service.Logger.Info("all the changesets have been synced to the new chain, closing the migrations listener.", log.Int("height", lastHeight))
+		return nil
+	} else {
+		currentHeight = lastHeight + 1
+	}
+
+	// Start polling the admin server for changesets
+	service.Logger.Info("start syncing changesets from old chain", log.Int("startHeight", cfg.StartHeight))
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
```

Comment:
Unless there's something else in this select, this select doesn't do anything useful.  Any blocking function call would get the context and be cancelled that way.  Even if we had good reason to do a context cancellation check prior to each iteration, we'd outdent the body of `default`.